src/rdo.rs

```diff
@@ -1114,24 +1114,55 @@ fn intra_frame_rdo_mode_decision<T: Pixel>(
   let num_modes_rdo: usize;
   let mut modes = ArrayVec::<[_; INTRA_MODES]>::new();
 
-  // If tx partition (i.e. fi.tx_mode_select) is enabled, don't use below intra prediction screening
-  if !fi.tx_mode_select {
-    let tx_size = bsize.tx_size();
-
-    // Reduce number of prediction modes at higher speed levels
-    num_modes_rdo = if (fi.frame_type == FrameType::KEY
+  // Reduce number of prediction modes at higher speed levels
+  num_modes_rdo = if (fi.frame_type == FrameType::KEY
+    && fi.config.speed_settings.prediction_modes
+      >= PredictionModesSetting::ComplexKeyframes)
+    || (fi.frame_type.has_inter()
       && fi.config.speed_settings.prediction_modes
-        >= PredictionModesSetting::ComplexKeyframes)
-      || (fi.frame_type.has_inter()
-        && fi.config.speed_settings.prediction_modes
-          >= PredictionModesSetting::ComplexAll)
-    {
-      7
+        >= PredictionModesSetting::ComplexAll)
+  {
+    7
+  } else {
+    3
+  };
+
+  let intra_mode_set = RAV1E_INTRA_MODES;
+
+  // Find mode with lowest rate cost
+  {
+    let mut z = 32768;
+    let probs_all = if fi.frame_type.has_inter() {
+      cw.get_cdf_intra_mode(bsize)
     } else {
-      3
-    };
+      cw.get_cdf_intra_mode_kf(tile_bo)
+    }
+    .iter()
+    .take(INTRA_MODES)
+    .map(|&a| {
+      let d = z - a;
+      z = a;
+      d
+    })
+    .collect::<ArrayVec<[_; INTRA_MODES]>>();
+
+    let mut probs = intra_mode_set
+      .iter()
+      .map(|&a| (a, probs_all[a as usize]))
+      .collect::<ArrayVec<[_; INTRA_MODES]>>();
+    probs.sort_by_key(|a| !a.1);
+
+    probs
+      .iter()
+      .take(num_modes_rdo / 2)
+      .for_each(|&(luma_mode, _prob)| modes.push(luma_mode));
```

Comment:
Not sure if [extend](https://doc.rust-lang.org/nightly/core/iter/trait.Extend.html) would be nicer that this, but it might :)